interfaces/prompting/patterns/variant.go

```diff
@@ -0,0 +1,426 @@
+				// Replace previous /**/* with /*/** before adding separator
+				components[len(components)-3] = component{compType: compSeparator}
+				components[len(components)-2] = component{compType: compGlobstar}
+				components[len(components)-1] = component{compType: compSeparatorDoublestar}
+			}
+			if !prevComponentsAre([]componentType{compSeparator}) {
+				// Collapse repeated separators
+				components = append(components, component{compType: compSeparator})
+			}
+		case '?':
+			reducePrevDoublestar()
+			consumeText()
+			if prevComponentsAre([]componentType{compGlobstar}) {
+				// Insert '?' before previous '*'
+				components[len(components)-1] = component{compType: compAnySingle}
+				components = append(components, component{compType: compGlobstar})
+			} else {
+				components = append(components, component{compType: compAnySingle})
+			}
+		case '⁑':
+			consumeText()
+			if prevComponentsAre([]componentType{compSeparatorDoublestar, compSeparator}) {
+				// Reduce /**/** to simply /** by removing most recent separator
+				components = components[:len(components)-1]
+			} else if prevComponentsAre([]componentType{compSeparator}) {
+				// Replace previous separator with separatorDoublestar
+				components[len(components)-1] = component{compType: compSeparatorDoublestar}
+			} else {
+				// Reduce to * since previous component is not a separator
+				addGlobstar()
+			}
+		case '*':
+			reducePrevDoublestar()
+			consumeText()
+			addGlobstar()
+		case '\\':
+			r2, _, err := rr.ReadRune()
+			if err != nil {
+				// Should be impossible, we just rendered this variant
+				return PatternVariant{}, errors.New(`internal error: trailing unescaped '\' character`)
+			}
+			switch r2 {
+			case '*', '?', '[', ']', '{', '}', '\\':
+				runes = append(runes, r, r2)
+			default:
+				// do not add r to runes if it's unnecessary
+				runes = append(runes, r2)
+			}
+		case '[', ']', '{', '}':
+			// Should be impossible, we just rendered this variant
+			return PatternVariant{}, fmt.Errorf(`internal error: unexpected unescaped '%v' character`, r)
+		default:
+			runes = append(runes, r)
+		}
+	}
+
+	consumeText()
+
+	if prevComponentsAre([]componentType{compSeparatorDoublestar, compSeparator, compGlobstar}) {
+		// If components end with /**/*, strip trailing /*
+		components = components[:len(components)-2]
+	}
+
+	// Add terminal marker or convert existing doublestar to terminal doublestar
+	if prevComponentsAre([]componentType{compSeparatorDoublestar, compSeparator}) {
+		components = components[:len(components)-2]
+		components = append(components, component{compType: compSeparatorDoublestarSeparatorTerminal})
+	} else if prevComponentsAre([]componentType{compSeparatorDoublestar}) {
+		components[len(components)-1] = component{compType: compSeparatorDoublestarTerminal}
+	} else {
+		components = append(components, component{compType: compTerminal})
+	}
```

Comment:
not sure if this isn't a tad clearer:

```suggestion
	components = append(components, component{compType: compTerminal})

	// various reductions around **
	switch {
	case prevComponentsAre([]componentType{compSeparatorDoublestar, compSeparator, compGlobstar, compTerminal}):
		// If components end with /**/*, reduce to terminal /**
		components = append(components[:len(components)-4], 
							component{compType: compSeparatorDoublestarSeparatorTerminal})
	case prevComponentsAre([]componentType{compSeparatorDoublestar, compSeparator, compTerminal}) {
		// reduce /**/ to a terminal /**
		components = append(components[:len(components)-3],
							component{compType: compSeparatorDoublestarSeparatorTerminal})
	case prevComponentsAre([]componentType{compSeparatorDoublestar, compTerminal}) {
		// reduce /**<terminal> to terminal /**
		components = append(components[:len(components)-2],
							component{compType: compSeparatorDoublestarSeparatorTerminal})
	}
```
or maybe not?

Reply:
Let's consider something like this for a follow-up PR... I'm undecided whether it's better to pre-add the terminal in all cases and then usually remove it later (treating `/**/*` as another terminal case, which I like), or to add either a terminal or a separator-doublestar-terminal depending on which is required (which I also like).